controllers/policies/admissionpolicy_gc_controller.go

```diff
@@ -0,0 +1,95 @@
+package policies
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	policiesv1alpha2 "github.com/kubewarden/kubewarden-controller/apis/policies/v1alpha2"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/admission"
+	"github.com/kubewarden/kubewarden-controller/internal/pkg/constants"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// AdmissionPolicyGCReconciler deletes namespaced policies that are no longer
+// attached to any policy server
+type AdmissionPolicyGCReconciler struct {
+	client.Client
+	Log        logr.Logger
+	Scheme     *runtime.Scheme
+	Reconciler admission.Reconciler
+}
+
+func (r *AdmissionPolicyGCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	admissionPolicy := policiesv1alpha2.AdmissionPolicy{}
+	if err := r.Get(ctx, req.NamespacedName, &admissionPolicy); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("cannot retrieve admission policy: %w", err)
+	}
+	policyServerWasDeleted := false
+	if admissionPolicy.Spec.PolicyServer != "" {
+		policyServer := policiesv1alpha2.PolicyServer{}
+		if err := r.Get(ctx, types.NamespacedName{Name: admissionPolicy.Spec.PolicyServer}, &policyServer); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, errors.Wrapf(err, "could not get policy server %s", admissionPolicy.Spec.PolicyServer)
+			}
+		} else {
+			policyServerWasDeleted = policyServer.DeletionTimestamp != nil
```

Comment:
Do we really need this check? as far as I know this will only be triggered when a `PolicyServer` is deleted, so this will always be true. Or am I missing something here?

Reply:
This is checking if the policy server was deleted in addition to the admission policy. In any case, when a controller starts it will trigger a `Reconcile` call for every resource that is targeting, so controllers whose nature is garbage collecting they just need to ignore all non-deleted resources.

Reply:
> But not delete admission policies that have been created whose policy server does not exist yet (they are unschedulable for example).

I don't see how a policy that does not have a PolicyServer can be reconciled in the gc_controller.
This will only be called when the `PolicyServer` for this `AdmissionPolicy` is being deleted, right? https://github.com/kubewarden/kubewarden-controller/pull/175/files#diff-cc3db0b7f2ac7bdcb785d4f604f3868cf4f02cfed967e7f43868921f2bf0ae47R76

Reply:
When the controller starts, an initial reconciliation loop is requested for all objects.